cardano-explorer-node/src/Explorer/Node.hs

```diff
@@ -392,6 +447,54 @@ localTxSubmissionCodec
 localTxSubmissionCodec =
   codecLocalTxSubmission nodeEncodeGenTx nodeDecodeGenTx Serialise.encode Serialise.decode
 
+-- the SingleTypeQueue can only contain ApplyBlock or RollBackToPoint objects, and never a mix of the 2
+-- Finish is in the same type as RollBackToPoint, and will also be exclusive
+-- when trying to insert a different one into the queue, you should first wait for the queue to return back to being Empty
+data Action = ApplyBlock ((ByronBlockOrEBB ByronConfig), BlockNo) | RollBackToPoint (Point (ByronBlockOrEBB ByronConfig)) | Finish
+data OnlyContains = Apply | Other | Empty deriving Eq
+data SingleTypeQueue = SingleTypeQueue
+  { stqContains :: TVar OnlyContains
+  , stqQueue :: TBQueue Action
+  }
+
+startDbThread :: Trace IO Text -> SingleTypeQueue -> Metrics -> IO ()
+startDbThread trce actionQueue metrics@Metrics{mQueuePre,mQueuePost} = do
+  (eActs, oldSize, newSize) <- atomically $ do
+    oldSize <- lengthTBQueue (stqQueue actionQueue)
+    contents <- readTVar (stqContains actionQueue)
+    eActs <- case contents of
+      Apply -> do
+        actions <- flushTBQueue $ stqQueue actionQueue
+        pure $ Right actions
+      Other -> do
+        action <- readTBQueue (stqQueue actionQueue)
+        pure $ Left action
+      Empty -> retry
+    newSize <- lengthTBQueue (stqQueue actionQueue)
+    if (newSize == 0)
+      then writeTVar (stqContains actionQueue) Empty
+      else pure ()
+    pure (eActs, oldSize, newSize)
+  Gauge.set (fromIntegral $ oldSize) mQueuePre
+  Gauge.set (fromIntegral $ newSize) mQueuePost
+  case eActs of
+    Right actions -> do
+      let
+        youSawNothing (ApplyBlock blk) = blk
```

Comment:
a better option, is maybe to have 2 queues, each of a single type, so i dont have to deal with the sum-type here, and then use the `OnlyContains` type to control which queue i read from